benchmarks/imagenet/resnet50/simclr.py

```diff
@@ -0,0 +1,100 @@
+from typing import List, Tuple
+
+from pytorch_lightning import LightningModule
+from torch import Tensor
+from torch.nn import Identity
+from torchvision.models import resnet50
+
+from lightly.loss.ntx_ent_loss import NTXentLoss
+from lightly.models.modules import SimCLRProjectionHead
+from lightly.models.utils import get_weight_decay_parameters
+from lightly.transforms import SimCLRTransform
+from lightly.utils.benchmarking import OnlineLinearClassifier
+from lightly.utils.lars import LARS
+from lightly.utils.scheduler import CosineWarmupScheduler
+
+
+class SimCLR(LightningModule):
+    def __init__(self, batch_size: int, num_classes: int) -> None:
+        super().__init__()
+        self.save_hyperparameters()
+        self.batch_size = batch_size
+
+        resnet = resnet50()
+        resnet.fc = Identity()  # Ignore classification head
+        self.backbone = resnet
+        self.projection_head = SimCLRProjectionHead()
+        self.criterion = NTXentLoss(temperature=0.1, gather_distributed=True)
+
+        self.online_classifier = OnlineLinearClassifier(num_classes=num_classes)
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.backbone(x)
+
+    def training_step(
+        self, batch: Tuple[List[Tensor], Tensor, List[str]], batch_idx: int
+    ) -> Tensor:
+        (view0, view1), targets = batch[0], batch[1]
+        features0 = self.forward(view0).flatten(start_dim=1)
+        features1 = self.forward(view1).flatten(start_dim=1)
+        z0 = self.projection_head(features0)
+        z1 = self.projection_head(features1)
+        loss = self.criterion(z0, z1)
+        self.log(
+            "train_loss", loss, prog_bar=True, sync_dist=True, batch_size=len(targets)
+        )
+
+        cls_loss, cls_log = self.online_classifier.training_step(
```

Comment:
The behavior here is not intuitive:
- Reading this code it's not clear what happens to the gradients from the online classification task
- Looking the the online classifier it's clear that the backpropagation is stopped
- I'd consider moving the `.detach` here

Reply:
I like that you went by composition over inheritance for the online classifier but it looks hacky to call a second training step from within the first one. Can we solve it by doing a forward pass through the online classifier and calculating the classification loss in the training step of SimCLR? I know we'd need to configure more optimizers etc but it's more transparent imo.

Reply:
> I'd consider moving the .detach here

Yep, good point.

> Can we solve it by doing a forward pass through the online classifier and calculating the classification loss in the training step of SimCLR? 

Then we also have to add the accuracy calculations and extra logging logic to every model which I would like to avoid.

> I know we'd need to configure more optimizers etc but it's more transparent imo.

Not sure I understand what you mean. Just note that we cannot use multiple different optimizers in a single training step without writing the optimization logic ourselves.